docs/0005-hashed-n-tuple-id-layout.md

```diff
@@ -0,0 +1,317 @@
+# OCFL Community Extension 0005: Hashed Truncated N-tuple Trees with Encapsulating Directory for OCFL Storage Hierarchies
+
+  * **Extension Name:** 0005-hashed-n-tuple-id-layout
+  * **Authors:** Ben Cail
+  * **Minimum OCFL Version:** 1.0
+  * **Obsoletes:** n/a
+  * **Obsoleted by:** n/a
+
+## Overview
+
+This storage root extension describes how to safely map OCFL object identifiers of any length, containing any characters, to OCFL object root directories.
+
+Using this extension, OCFL object identifiers are hashed and encoded as hex strings (all letters lower-case). These digests are then divided into _N_ n-tuple segments, which are used to create nested paths under the OCFL storage root. Finally, the OCFL object identifier is percent-encoded to create a directory name for the OCFL object root (see Encapsulation Directory section below).
+
+The n-tuple segments approach allows OCFL object identifiers to be evenly distributed across the storage hierarchy. The maximum number of files under any given directory is controlled by the number of characters in each n-tuple, and the tree depth is controlled by the number of n-tuple segments each digest is divided into. The encoded encapsulation directory name provides visibility into the object identifier from the file path (see "Encapsulation Directory" section below for details.
+
+## Encapsulation Directory
+
+For basic OCFL object identifiers, the object identifier is used as the name of the encapsulation directory (ie. the object root directory).
+
+Some object identifiers could contain characters that are not safe for directory names on all filesystems. When an unsafe character is encountered in an object identifier, it is percent-encoded using the lower-case hex characters of its UTF-8 bytes.
+
+Some object identifiers could also result in an encoded string that is longer than 255 characters, so it would not be safe to use as a directory name. In that case, the percent-encoded object identifier is truncated to 100 characters, and then the digest of the original object identifier is appended to the encoded object identifier like this: <encoded-object-identifier-first-100-chars>-<digest>.
```

Comment:
There is the possibility for a collision if you only truncate ids that are longer than 255 characters. Take you're example, the ids `abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghij` and `abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghij-55b432806f4e270da0cf23815ed338742179002153cd8d896f23b3e2d8a14359` would both be mapped to `abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghij-55b432806f4e270da0cf23815ed338742179002153cd8d896f23b3e2d8a14359`.

Granted, this collision is unlikely, and even if both of those ids were used, the collision would only occur if the n-tuples were also the same or a flat layout was used.

That said, I still think that it is better to encode all ids that are longer than the length at which long ids are truncated to. That is to say, truncate any id longer than 100 characters. You could increase the size, but only by a little. sha512 digests are 128 characters, which means that there are only 126 characters remaining (an additional character is used for the `-`). So, the largest you could get away with is truncated every id longer than 126 characters to 126 characters and then appending the digest.

Another possibility is that you could base the point at which you truncate on the digest algorithm that's used. So, with sha256 you could truncate at 190, for example.

Reply:
Thanks - I'll fix that.